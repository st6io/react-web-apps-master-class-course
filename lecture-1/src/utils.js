import { get } from 'lodash';

export const currency = {
  usd: 'USD',
  bgn: 'BGN',
  eur: 'EUR',
};

export const EUR_TO_BGN_RATE = 1.95583;

const random = (min, max) => Math.random() * (max - min) + min;

const getCurrencies = () => {
  const usdToBgn = random(1.6564, 1.7923, 4);
  const usdToEur = usdToBgn / 1.95583;

  return {
    [currency.usd]: {
      [currency.bgn]: usdToBgn,
      [currency.eur]: usdToEur,
      [currency.usd]: 1,
    },
    [currency.bgn]: {
      [currency.usd]: 1 / usdToBgn,
      [currency.eur]: 1 / EUR_TO_BGN_RATE,
      [currency.bgn]: 1,
    },
    [currency.eur]: {
      [currency.usd]: 1 / usdToEur,
      [currency.bgn]: EUR_TO_BGN_RATE,
      [currency.eur]: 1,
    },
  };
};

const currentCurrencies = getCurrencies();

export const toFixed = (value, precision = 2) => {
  const f = 10 ** precision;

  return Math.round(value * f) / f;
};

export const convertCurrency = (
  value,
  currency,
  targetCurrency,
  precision = 2,
) => {
  const input = parseFloat(value);
  if (Number.isNaN(input)) {
    return '';
  }

  const ratio = get(currentCurrencies, `${currency}.${targetCurrency}`, NaN);
  if (Number.isNaN(ratio)) {
    throw new Error(`cannot convert ${currency} to ${targetCurrency}`);
  }

  const converted = value * ratio;

  return toFixed(converted, precision);
};

const delay = (result, timeout = 250) =>
  new Promise(resolve => setTimeout(() => resolve(result), timeout));

export const fetchTempertature = () => delay('20C', 1000);

export const fetchHumidity = () => delay('54%');
