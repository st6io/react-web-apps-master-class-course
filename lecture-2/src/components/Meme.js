import React from 'react';
import styles from './Meme.module.scss';

const Meme = ({ topText, bottomText, image }) => {
  return (
    <div className={styles.meme}>
      <div className={styles.topText}>{topText}</div>
      <div className={styles.bottomText}>{bottomText}</div>
      <img src={image} alt="meme" className={styles.memeImage} />
    </div>
  );
};

export default Meme;
