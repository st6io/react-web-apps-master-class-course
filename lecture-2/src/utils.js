const images = [
  'https://www.maxim.com/.image/t_share/MTYwNjI1MDQ4NTM1NzA0NzQ3/jocko-willink-attack-mondays.png',
  'https://i.ytimg.com/vi/ljqra3BcqWM/maxresdefault.jpg',
  'https://peterattiamd.com/wp-content/uploads/2019/06/jocko-part2-2x1.png',
  'https://i.ytimg.com/vi/Ls5yHTGVzgU/maxresdefault.jpg',
  'https://miro.medium.com/max/3600/0*uvDOX4ea2mGXdAQ9',
];

export const getRandomImage = async () => {
  const response = await fetch(
    'https://qrng.anu.edu.au/API/jsonI.php?length=1&type=uint16',
  );

  const {
    data: [rnd],
  } = await response.json();

  return images[rnd % images.length];
};

export const sum = (a, b) => a + b;
