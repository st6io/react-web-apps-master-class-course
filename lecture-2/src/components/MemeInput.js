import React from 'react';
import Row from './Row';

import styles from './MemeInput.module.scss';
import { getRandomImage } from '../utils';

export default class MemeInput extends React.Component {
  state = {
    showUrl: false,
  };

  toggleUrlInput = () => {
    this.setState(({ showUrl }) => ({ showUrl: !showUrl }));
  };

  onTextChange = ({ target: { name, value } }) => {
    this.props.onInputChange({ [name]: value });
  };

  shuffleImage = async () => {
    const image = await getRandomImage();

    this.props.onImageChange(image);
  };

  render() {
    const { topText, bottomText, url, className } = this.props;

    return (
      <div className={className}>
        <Row>
          <input
            type="text"
            value={topText}
            onChange={this.onTextChange}
            name="topText"
          />
        </Row>
        <Row>
          <input
            type="text"
            value={bottomText}
            onChange={this.onTextChange}
            name="bottomText"
          />
        </Row>

        <Row>
          <div className={styles.leftColumn}>
            <button onClick={this.toggleUrlInput}>Via Url</button>
            <button id="shuffle" onClick={this.shuffleImage}>
              Shuffle
            </button>
          </div>
        </Row>

        {this.state.showUrl && (
          <Row>
            <input
              type="text"
              value={url}
              onChange={this.onTextChange}
              name="url"
            />
          </Row>
        )}
      </div>
    );
  }
}
