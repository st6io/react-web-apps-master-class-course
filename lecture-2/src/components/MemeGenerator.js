import React from 'react';

import Container from './Container';
import Card from './Card';
import styles from './MemeGenerator.module.scss';
import MemeInput from './MemeInput';
import Meme from './Meme';

const defaultImage =
  'https://miro.medium.com/max/1200/1*y6C4nSvy2Woe0m7bWEn4BA.png';

export default class MemeGenerator extends React.Component {
  state = {
    topText: 'React',
    bottomText: 'is awesome!',
    url: '',
    image: defaultImage,
  };

  handleInputChange = input => {
    this.setState(input);
  };

  handleImageChange = image => {
    this.setState({ image });
  };

  render() {
    const { topText, bottomText, url, image } = this.state;

    return (
      <Container className={styles.container}>
        <Card className={styles.card}>
          <MemeInput
            {...{ topText, bottomText, url }}
            onInputChange={this.handleInputChange}
            onImageChange={this.handleImageChange}
          />
          <Meme {...{ topText, bottomText, image }} />
        </Card>
      </Container>
    );
  }
}
