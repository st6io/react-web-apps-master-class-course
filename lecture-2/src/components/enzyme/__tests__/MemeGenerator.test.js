import React from 'react';
import { shallow, mount } from 'enzyme';
import MemeGenerator from '../../MemeGenerator';
import MemeInput from '../../MemeInput';
import Meme from '../../Meme';
import { getRandomImage } from '../../../utils';

jest.mock('../../../utils', () => ({
  getRandomImage: jest.fn(),
}));

describe('MemeGenerator Component', () => {
  it('smoke test', () => {
    expect(shallow(<MemeGenerator />)).toExist();
  });

  it('should match snapshot', () => {
    const wrapper = shallow(<MemeGenerator />);

    expect(wrapper).toMatchSnapshot();
  });

  it('full rendering should match snapshot', () => {
    const wrapper = mount(<MemeGenerator />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should match snapshot when topText changes', () => {
    const wrapper = shallow(<MemeGenerator />);
    const expectedTopText = 'expected top text';

    wrapper
      .find(MemeInput)
      .simulate('inputChange', { topText: expectedTopText });

    const memeWrapper = wrapper.find(Meme);

    expect(memeWrapper).toHaveProp('topText', expectedTopText);
  });

  it('should match snapshot when image changes', () => {
    const wrapper = shallow(<MemeGenerator />);
    const expectedImage = 'https://test.image/png';

    wrapper.find(MemeInput).simulate('imageChange', expectedImage);

    const memeWrapper = wrapper.find(Meme);

    expect(memeWrapper).toHaveProp('image', expectedImage);
  });

  it('test shuffle integration', async () => {
    const wrapper = mount(<MemeGenerator />);
    const expectedImage = 'https://test.img/jpg';

    const resolveExpectedImage = Promise.resolve(expectedImage);
    getRandomImage.mockImplementation(() => resolveExpectedImage);

    wrapper.find('#shuffle').simulate('click');

    await resolveExpectedImage;
    wrapper.update();

    expect(wrapper.find('img')).toHaveProp('src', expectedImage);
  });
});
