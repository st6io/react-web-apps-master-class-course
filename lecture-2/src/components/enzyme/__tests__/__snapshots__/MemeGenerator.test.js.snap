// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemeGenerator Component full rendering should match snapshot 1`] = `
<MemeGenerator>
  <Container
    className="container"
  >
    <div
      className="container container"
    >
      <Card
        className="card"
      >
        <div
          className="card card"
        >
          <MemeInput
            bottomText="is awesome!"
            onImageChange={[Function]}
            onInputChange={[Function]}
            topText="React"
            url=""
          >
            <div>
              <Row>
                <div
                  className="row"
                >
                  <input
                    name="topText"
                    onChange={[Function]}
                    type="text"
                    value="React"
                  />
                </div>
              </Row>
              <Row>
                <div
                  className="row"
                >
                  <input
                    name="bottomText"
                    onChange={[Function]}
                    type="text"
                    value="is awesome!"
                  />
                </div>
              </Row>
              <Row>
                <div
                  className="row"
                >
                  <div
                    className="leftColumn"
                  >
                    <button
                      onClick={[Function]}
                    >
                      Via Url
                    </button>
                    <button
                      id="shuffle"
                      onClick={[Function]}
                    >
                      Shuffle
                    </button>
                  </div>
                </div>
              </Row>
            </div>
          </MemeInput>
          <Meme
            bottomText="is awesome!"
            image="https://miro.medium.com/max/1200/1*y6C4nSvy2Woe0m7bWEn4BA.png"
            topText="React"
          >
            <div
              className="meme"
            >
              <div
                className="topText"
              >
                React
              </div>
              <div
                className="bottomText"
              >
                is awesome!
              </div>
              <img
                alt="meme"
                className="memeImage"
                src="https://miro.medium.com/max/1200/1*y6C4nSvy2Woe0m7bWEn4BA.png"
              />
            </div>
          </Meme>
        </div>
      </Card>
    </div>
  </Container>
</MemeGenerator>
`;

exports[`MemeGenerator Component should match snapshot 1`] = `
<Container
  className="container"
>
  <Card
    className="card"
  >
    <MemeInput
      bottomText="is awesome!"
      onImageChange={[Function]}
      onInputChange={[Function]}
      topText="React"
      url=""
    />
    <Meme
      bottomText="is awesome!"
      image="https://miro.medium.com/max/1200/1*y6C4nSvy2Woe0m7bWEn4BA.png"
      topText="React"
    />
  </Card>
</Container>
`;
